debian/upgrade-utils/desktop-grid/meta-get.1
.\" Man Page for meta-get
.TH "meta-get" "1"
.SH "NAME"
meta-get \- utils for get file meta info.
.SH "SYNOPSIS"
.B meta-get [OPTION] <path> [keys..]
.SH "DESCRIPTION"
meta-get
This program provides several commands for get file meta info.
.TP
\fB -t, --tree\fR
Tree
.TP
\fB -f, --file\fR
Tree file
.TP
\fB -r, --recursive\fR
Recursive
.SH "BUGS"
.SS Should you encounter any bugs, they may be reported at: 
https://gitee.com/openkylin/peony/issues
.SH "AUTHORS"
.SS This Manual Page has been written for the UKUI Desktop Environment by:
Yue Lan <lanyue@kylinos.cn> (2024)


debian/changelog
peony (4.0.0.1-1) unstable; urgency=medium

  * Disable the auto partition mount to fix UbuntuKylin install error.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Wed, 10 Apr 2024 18:43:49 +0800

peony (4.0.0.0-1.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Rename libraries for 64-bit time_t transition.  Closes: #1063217

 -- Benjamin Drung <bdrung@debian.org>  Thu, 29 Feb 2024 20:19:01 +0000

peony (4.0.0.0-1) unstable; urgency=medium

  * New upstream release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 23 Jan 2024 10:42:22 +0800

peony (3.2.4-1) unstable; urgency=medium

  * New upstream release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 22 Feb 2022 16:11:28 +0800

peony (3.2.3-1) unstable; urgency=medium

  * New upstream release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 17 Aug 2021 19:35:54 +0800

peony (3.0.6-1) unstable; urgency=medium

  * New upstream release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Sat, 28 Nov 2020 15:53:31 +0800


peony (3.0.5-1) unstable; urgency=medium

  * New upstream bugfix release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Sun, 27 Sep 2020 19:28:06 +0800

peony (3.0.4-1) unstable; urgency=medium

  * New upstream bugfix release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Sat, 26 Sep 2020 20:26:06 +0800

peony (3.0.3-1) unstable; urgency=medium

  * New upstream bugfix release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 07 Sep 2020 16:16:47 +0800

peony (3.0.2-2) unstable; urgency=medium

  * debian/control:
    - libpeony3: add breaks with libpeony2. (Closes: #968595)

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 18 Aug 2020 13:11:28 +0000

peony (3.0.2-1) unstable; urgency=medium

  * New upstream release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Wed, 12 Aug 2020 09:22:02 +0000

peony (3.0.1-1) unstable; urgency=medium

  * New upstream release. (Closes: #957676)
  * debian:
    - drop libgtk2.0-dev and qt5-gtk2-platformtheme.
    - update the install files.
    - libpeony2 -> libpeony3.
    - Update watch file.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Sat, 08 Aug 2020 14:22:20 +0800

peony (2.1.2-1) unstable; urgency=medium

  * New upstream bugfix release.
    - Fix enter key as shortcut issues such as open selected file.
    - Fix that file in drag state not show file name.
    - Fix that file rename as empty and rename file change current sort.
    - Fix that desktop app link show incorrect icon.
    - Fix generate file thumbnail issue and ensure thread safety.
    - Fix that screen mirror not show view.
    - Fix file handle error.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 06 Apr 2020 19:57:13 +0800

peony (2.1.1-1) unstable; urgency=medium

  * New upstream bugfix release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 30 Mar 2020 21:42:00 +0800

peony (2.1.0-1) unstable; urgency=medium

  * New upstream release.
    - Use custom header bar, and border rounded window.
    - Remove current advance search page, and relpace it with file label
      page.
    - Improve the thumbnail frameworks.
    - Improve the mouse right click interaction in desktop view.
    - Allow unrelated trash or delete file operation executed when a
      file operation is running.
    - Refactor the desktop window background displaying process.
  * debian/libpeony3.symbols:
    - update symbols file.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 23 Mar 2020 17:56:24 +0800

peony (2.0.1-2) unstable; urgency=medium

  * debian/libpeony3.symbols:
    - fixed symbols for armel.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Thu, 20 Feb 2020 11:00:36 +0800

peony (2.0.1-1) unstable; urgency=medium

  * New upstream release.
  * debian/control:
    - Add build-depends pkg-kde-tools.
  * debian/rule:
    - using pkgkde_symbolshelper.
  * debian/libpeony3.symbols.
    - Using pkg-kde-tools update the symbols file. (Closes: #951607)

 -- handsome_feng <jianfengli@ubuntukylin.com>  Wed, 19 Feb 2020 14:15:38 +0800

peony (2.0.0-1) unstable; urgency=medium

  * New upstream release.
    - Refactor peony ui with Qt (from Gtk3).
  * debian/control:
    - Bump to compat 12.
    - Bump Standards-Version to 4.5.0.
    - Refresh the dependency.
    - Set Rules-requires-Root to no.
  * debian/copyright: Refreash the copyright entry.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 04 Feb 2020 11:09:14 +0800

peony (1.1.7-1) unstable; urgency=medium

  * Fix that the outline of the file name label is incomplete.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Wed, 16 Oct 2019 16:44:58 +0800

peony (1.1.6-1) unstable; urgency=medium

  * Add file path to the address bar that can click.
  * Add drag directories to favorites and delete favorites directories
    function.
  * Add folder image and set location button size.
  * Add search entry delete function.
  * Modify sidebar, statusbar, locationbar style.
  * Display picture according to its own width and height.
  * Update zh_TW.po and zh_CN.po.
  * Fix trash restore files error.
  * Fix that computer, recent and trash is English in addressbar.
  * Fix the location bar error.
  * Reload window after desktop-is-home-dir changed. (LP: #1831857)
  * Fix the crash when using listview mode. (LP: #1827994)
  * debian/control:
    - Drop engrampa from suggests.
    - Fix symbols-file-missing-build-depends-package-field.
    - Bump Standards-Version to 4.4.0.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 06 Aug 2019 14:51:41 +0800

peony (1.1.5-1) unstable; urgency=medium

  * Add the function of preview files.
  * Add format function.
  * Change the name style when copy files.
  * Simplify disk label description.
  * Update es.po, fr.po, pt.po, ru.po.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Thu, 27 Dec 2018 20:43:29 +0800

peony (1.1.4-1) unstable; urgency=medium

  * New upstream release

 -- Aron Xu <aron@debian.org>  Mon, 22 Oct 2018 23:41:44 +0800

peony (1.1.3-1) unstable; urgency=medium

debian/upgrade-utils/desktop-grid/meta-get.c
//#include "config.h"
#include "metatree.h"
#include <glib/gstdio.h>

static char *treename = NULL;
static char *treefilename = NULL;
static gboolean recursive = FALSE;
static GOptionEntry entries[] =
{
  { "tree", 't', 0, G_OPTION_ARG_STRING, &treename, "Tree", NULL},
  { "file", 'f', 0, G_OPTION_ARG_STRING, &treefilename, "Tree file", NULL},
  { "recursive", 'r', 0, G_OPTION_ARG_NONE, &recursive, "Recursive", NULL},
  { NULL }
};

static gboolean
print_key (const char *key,
	   MetaKeyType type,
	   gpointer value,
	   gpointer user_data)
{
  int indent = GPOINTER_TO_INT (user_data);
  char **values;
  int i;

  g_assert (type != META_KEY_TYPE_NONE);

  if (type == META_KEY_TYPE_STRING)
    g_print ("%*s%s=%s\n", indent, "",key, (char *)value);
  else
    {
      values = value;
      g_print ("%*s%s=[", indent, "",key);
      for (i = 0; values[i] != NULL; i++)
	{
	  if (values[i+1] != NULL)
	    g_print ("%s,", values[i]);
	  else
	    g_print ("%s", values[i]);
	}
      g_print ("]\n");
    }
  return TRUE;
}

static gboolean
prepend_name (const char *entry,
	      guint64 last_changed,
	      gboolean has_children,
	      gboolean has_data,
	      gpointer user_data)
{
  GList **children = user_data;

  *children = g_list_prepend (*children,
			      g_strdup (entry));
  return TRUE;
}

static void
enum_keys (MetaTree *tree, char *path,
	   gboolean recurse, int indent)
{
  GList *children, *l;
  char *child_name, *child_path;

  g_print ("%*s%s\n", indent, "", path);
  meta_tree_enumerate_keys (tree, path,
			    print_key, GINT_TO_POINTER (indent+1));

  if (recurse)
    {
      children = NULL;
      meta_tree_enumerate_dir (tree, path,
			       prepend_name,
			       &children);
      for (l = children; l != NULL; l = l->next)
	{
	  child_name = l->data;
	  child_path = g_build_filename (path, l->data, NULL);
	  g_free (child_name);

	  enum_keys (tree, child_path, recurse, indent + 3);

	  g_free (child_path);
	}
      g_list_free (children);
    }
}

int
main (int argc,
      char *argv[])
{
  MetaTree *tree;
  GError *error = NULL;
  GOptionContext *context;
  MetaKeyType type;
  const char *path, *key;
  MetaLookupCache *lookup;
  struct stat statbuf;
  char *tree_path;
  char **strings;
  int i, j;

  context = g_option_context_new ("<path> [keys..]- read metadata");
  g_option_context_add_main_entries (context, entries, NULL);
  if (!g_option_context_parse (context, &argc, &argv, &error))
    {
      g_printerr ("option parsing failed: %s\n", error->message);
      return 1;
    }

  if (argc < 2)
    {
      g_printerr ("no path specified\n");
      return 1;
    }
  path = argv[1];

  if (treefilename)
    {
      tree = meta_tree_open (treefilename, FALSE);
      if (tree)
	tree_path = g_strdup (path);

      if (tree == NULL)
	{
	  g_printerr ("can't open metadata file %s\n", treefilename);
	  return 1;
	}
    }
  else if (treename)
    {
      tree = meta_tree_lookup_by_name (treename, FALSE);
      if (tree)
	tree_path = g_strdup (path);

      if (tree == NULL)
	{
	  g_printerr ("can't open metadata tree %s\n", path);
	  return 1;
	}
    }
  else
    {
      lookup = meta_lookup_cache_new ();
      if (g_lstat (path, &statbuf) != 0)
	{
	  g_printerr ("can't find file %s\n", path);
	  return 1;
	}
      tree = meta_lookup_cache_lookup_path (lookup,
					    path,
					    statbuf.st_dev,
					    FALSE,
					    &tree_path);
      meta_lookup_cache_free (lookup);

      if (tree == NULL)
	{
	  g_printerr ("can't open metadata tree for file %s\n", path);
	  return 1;
	}
    }

  if (argc > 2)
    {
      for (i = 2; i < argc; i++)
	{
	  key = argv[i];
	  type = meta_tree_lookup_key_type  (tree, tree_path, key);
	  if (type == META_KEY_TYPE_NONE)
	    g_print ("%s Not set\n", key);
	  else if (type == META_KEY_TYPE_STRING) {
	    g_print ("%s\n", meta_tree_lookup_string (tree, path, key));
	  }
	  else
	    {
	      strings = meta_tree_lookup_stringv (tree, path, key);
	      g_print ("%s=[", key);
	      for (j = 0; strings[j] != NULL; j++)
		{
		  if (strings[j+1] == NULL)
		    g_print ("%s", strings[j]);
		  else
		    g_print ("%s,", strings[j]);
		}
	      g_print ("]\n");
	    }
	}
    }
  else
    {
      enum_keys (tree, tree_path, recursive, 0);
    }

  return 0;
}
